File: examples/learn/topic/protobuf/python/protobuf_node_creator.py
# Copyright (c) 2019 The Felicia Authors. All rights reserved.
# Use of this source code is governed by a BSD-style license that can be
# found in the LICENSE file.

import sys

import import_order_resolver  # Should be before import felicia_py
import felicia_py as fel
from felicia.core.protobuf.master_data_pb2 import NodeInfo
import felicia_py.command_line_interface as cli

from examples.learn.topic.topic_create_flag import TopicCreateFlag
from protobuf_publishing_node import ProtobufPublishingNode
from protobuf_subscribing_node import ProtobufSubscribingNode


def main():
    fel.felicia_init()

    delegate = TopicCreateFlag()
    parser = cli.FlagParser()
    parser.set_program_name('protobuf_node_creator')
    if not parser.parse(len(sys.argv), sys.argv, delegate):
        sys.exit(1)

    master_proxy = fel.master_proxy
    s = master_proxy.start()
    if not s.ok():
        print("{} {}.".format(cli.RED_ERROR, s), file=sys.stderr)
        sys.exit(1)

    node_info = NodeInfo()
    node_info.name = delegate.name_flag.value

    ssl_server_context = None
    if delegate.is_publshing_node_flag.value:
        if delegate.use_ssl_flag.value:
            cert_file_path = fel.FilePath("examples/cert/server.crt")
            private_key_file_path = fel.FilePath(
                "examples/cert/server.key")
            ssl_server_context = fel.channel.SSLServerContext.new_ssl_server_context(
                cert_file_path,
                private_key_file_path
            )
        master_proxy.request_register_node(
            ProtobufPublishingNode, node_info, delegate, ssl_server_context)
    else:
        master_proxy.request_register_node(
            ProtobufSubscribingNode, node_info, delegate)

    fel.main_thread.run()


if __name__ == '__main__':
    main()

File: examples/learn/topic/protobuf/cc/protobuf_node_creator.cc
// Copyright (c) 2019 The Felicia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "felicia/core/channel/socket/ssl_server_context.h"
#include "felicia/core/felicia_init.h"
#include "felicia/core/master/master_proxy.h"
#include "felicia/core/thread/main_thread.h"

#include "examples/learn/topic/protobuf/cc/protobuf_publishing_node.h"
#include "examples/learn/topic/protobuf/cc/protobuf_subscribing_node.h"
#include "examples/learn/topic/topic_create_flag.h"

namespace felicia {

int RealMain(int argc, char* argv[]) {
  FeliciaInit();

  TopicCreateFlag delegate;
  FlagParser parser;
  parser.set_program_name("protobuf_node_creator");
  if (!parser.Parse(argc, argv, &delegate)) {
    return 1;
  }

  MasterProxy& master_proxy = MasterProxy::GetInstance();
  Status s = master_proxy.Start();
  if (!s.ok()) {
    std::cerr << kRedError << s << std::endl;
    return 1;
  }

  NodeInfo node_info;
  node_info.set_name(delegate.name_flag()->value());

  std::unique_ptr<SSLServerContext> ssl_server_context;
  if (delegate.is_publishing_node_flag()->value()) {
    if (delegate.use_ssl_flag()->value()) {
      base::FilePath cert_file_path =
          base::FilePath(FILE_PATH_LITERAL("examples/cert/server.crt"));
      base::FilePath private_key_file_path =
          base::FilePath(FILE_PATH_LITERAL("examples/cert/server.key"));
      ssl_server_context = SSLServerContext::NewSSLServerContext(
          cert_file_path, private_key_file_path);
    }
    master_proxy.RequestRegisterNode<ProtobufPublishingNode>(
        node_info, delegate, ssl_server_context.get());
  } else {
    master_proxy.RequestRegisterNode<ProtobufSubscribingNode>(node_info,
                                                              delegate);
  }

  MainThread& main_thread = MainThread::GetInstance();
  main_thread.Run();

  return 0;
}

}  // namespace felicia

int main(int argc, char* argv[]) { return felicia::RealMain(argc, argv); }
